server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterAllocationSimulationTests.java

```diff
@@ -0,0 +1,549 @@
+import org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.snapshots.SnapshotShardSizeInfo;
+import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.test.ClusterServiceUtils;
+import org.elasticsearch.test.gateway.TestGatewayAllocator;
+import org.elasticsearch.test.junit.annotations.TestLogging;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+import static org.elasticsearch.cluster.node.DiscoveryNodeRole.DATA_COLD_NODE_ROLE;
+import static org.elasticsearch.cluster.node.DiscoveryNodeRole.DATA_HOT_NODE_ROLE;
+import static org.elasticsearch.cluster.node.DiscoveryNodeRole.DATA_WARM_NODE_ROLE;
+import static org.elasticsearch.cluster.node.DiscoveryNodeRole.MASTER_ROLE;
+
+@TestLogging(reason = "progress reporting", value = "org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceComputer:DEBUG")
+public class ClusterAllocationSimulationTests extends ESAllocationTestCase {
+
+    private static final Logger logger = LogManager.getLogger(ClusterAllocationSimulationTests.class);
+
+    private static int nodeCount(long tierBytes, long nodeBytes, double tierWriteLoad, double nodeIndexingThreads, int minCount) {
+        var watermarkNodeBytes = (long) (nodeBytes * 0.85);
+        return Math.max(
+            Math.toIntExact((tierBytes + watermarkNodeBytes - 1) / watermarkNodeBytes),
+            Math.max((int) Math.ceil(tierWriteLoad / nodeIndexingThreads), minCount)
+        ) + between(0, 2);
+    }
+
+    public void testBalanceQuality() throws IOException {
+
+        final var shardSizesByIndex = new HashMap<String, Long>();
+        final var tiers = new String[] { DataTier.DATA_HOT, DataTier.DATA_WARM, DataTier.DATA_COLD };
+        final var tierSizes = Arrays.stream(tiers).collect(Collectors.toMap(Function.identity(), s -> 0L));
+        final var tierWriteLoads = Arrays.stream(tiers).collect(Collectors.toMap(Function.identity(), s -> 0.0));
+        final var maxSizeVariance = ByteSizeValue.ofMb(20).getBytes();
+
+        final var metadataBuilder = Metadata.builder();
+        for (var dataStreamIndex = between(5, 10); dataStreamIndex >= 0; dataStreamIndex -= 1) {
+            final var hotIndices = between(2, 10);
+            final var warmIndices = between(10, 20);
+            final var coldIndices = between(10, 40);
+            final var hotShards = randomFrom(1, 2, 4, 8);
+            final var shrunkShards = randomFrom(1, 2);
```

Comment:
This still looks sizeable. What is the average run time for this test?

Reply:
I think we should keep those large scenarios to ensure that we converge in a reasonable time?

Reply:
On my machine these runs almost always complete in under a couple of seconds. Sometimes a little longer, but the distribution looks fine for this kind of test.

I think if we go for much larger scenarios we won't get much better coverage, we'll just burn more CI time.